File: ui/src/lib/events.ts
import { listen } from "@tauri-apps/api/event";

// Mirrors brail_core::events::AppEvent's #[serde(tag = "type", content = "data")]
// representation: every event arrives as { type: "...", data: {...} }.
export type AppEvent =
  | { type: "HardwareDetected"; data: unknown }
  | { type: "RecordingStarted"; data: { output_path: string } }
  | { type: "RecordingStopped"; data: { output_path: string; final_size_bytes: number } }
  | { type: "RecordingPaused"; data: null }
  | { type: "RecordingResumed"; data: null }
  | { type: "RecordingStatsUpdated"; data: { duration_secs: number; file_size_bytes: number; disk_write_mb_per_sec: number } }
  | { type: "RecordingRecovered"; data: { recovered_path: string } }
  | { type: "StreamConnecting"; data: null }
  | { type: "StreamConnected"; data: null }
  | { type: "StreamDisconnected"; data: { reason: string } }
  | { type: "StreamReconnecting"; data: { attempt: number; max_attempts: number } }
  | {
      type: "StreamStatsUpdated";
      data: {
        connection_state: "Idle" | "Connecting" | "Connected" | "Reconnecting" | "Failed";
        duration_secs: number;
        upload_bitrate_kbps: number;
        target_bitrate_kbps: number;
        dropped_frames: number;
        dropped_frames_percent: number;
      };
    }
  | { type: "ReplaySaved"; data: { output_path: string } }
  | { type: "ScreenshotSaved"; data: { output_path: string } }
  | {
      type: "AudioLevels";
      data: { microphone_rms: number; desktop_rms: number; microphone_peak: number; desktop_peak: number };
    }
  | {
      type: "AdaptiveRecommendation";
      data: { action: string; reason: string; severity: string; auto_applied: boolean };
    }
  | { type: "BenchmarkComplete"; data: { summary: string } }
  | { type: "CaptureStatsUpdated"; data: { capture_fps: number; frames_captured: number; frames_dropped_capture: number } }
  | {
      type: "ResourceStatsUpdated";
      data: {
        process_ram_mb: number;
        process_cpu_percent: number;
        gpu_usage_percent: number | null;
        gpu_vram_used_mb: number | null;
      };
    }
  | { type: "CaptureSourceLost"; data: { source_name: string } }
  | { type: "Warning"; data: { code: string; message: string } }
  | { type: "Error"; data: unknown };

export function onAppEvent(handler: (event: AppEvent) => void): Promise<() => void> {
  return listen<AppEvent>("brail://event", (e) => handler(e.payload));
}

File: crates/brail-core/src/events.rs
use serde::{Deserialize, Serialize};

use crate::error::BrailError;
use crate::stats::{CaptureStats, EncodeStats, RecordingStats, ResourceStats, StreamStats};

/// Everything the backend pipeline can tell the UI. Emitted over a single
/// Tauri event channel (`brail://event`) so the frontend has one listener
/// instead of a dozen ad-hoc channels — see src-tauri/src/events.rs for the
/// emit side and ui/src/lib/events.ts for the listener.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "data")]
pub enum AppEvent {
    HardwareDetected(crate::profile::CapabilityProfile),

    RecordingStarted { output_path: String },
    RecordingStopped { output_path: String, final_size_bytes: u64 },
    RecordingPaused,
    RecordingResumed,
    RecordingStatsUpdated(RecordingStats),
    RecordingRecovered { recovered_path: String },

    StreamConnecting,
    StreamConnected,
    StreamDisconnected { reason: String },
    StreamReconnecting { attempt: u32, max_attempts: u32 },
    StreamStatsUpdated(StreamStats),
    StreamTestResult { success: bool, message: String, round_trip_ms: Option<f64> },

    ReplaySaved { output_path: String },
    ScreenshotSaved { output_path: String },
    ReplayBufferFilled,

    CaptureStatsUpdated(CaptureStats),
    EncodeStatsUpdated(EncodeStats),
    ResourceStatsUpdated(ResourceStats),

    CaptureSourceLost { source_name: String },
    MonitorConfigurationChanged,

    /// Live audio meter levels, batched at the same 1 Hz cadence as the
    /// other stats so the meters never drive high-frequency UI renders.
    AudioLevels { microphone_rms: f32, desktop_rms: f32, microphone_peak: f32, desktop_peak: f32 },

    /// A recommendation from the Brail Adaptive Engine. Emitted whether or
    /// not it was auto-applied, so the user is always told (§13: never
    /// silently destroy quality).
    AdaptiveRecommendation { action: String, reason: String, severity: String, auto_applied: bool },

    BenchmarkComplete { summary: String },

    Warning { code: String, message: String },
    Error(BrailError),
}
